src/peer_select.cc

```diff
@@ -564,17 +564,16 @@ PeerSelector::selectPinned()
     if (!request->pinnedConnection())
         return;
 
-    if (Comm::IsConnOpen(request->pinnedConnection()->validatePinnedConnection(request))) {
-        const auto pear = request->pinnedConnection()->pinnedPeer();
-        const bool usePinned = pear ? peerAllowedToUse(pear, this) : (direct != DIRECT_NO);
-        if (usePinned) {
-            addSelection(pear, PINNED);
-            if (entry)
-                entry->ping_status = PING_DONE; // skip ICP
-        }
-    }
-    // If the pinned connection is prohibited (for this request) or gone, then
+    const auto pear = request->pinnedConnection()->pinnedPeer();
+    const bool usePinned = pear ? peerAllowedToUse(pear, this) : (direct != DIRECT_NO);
+    // If the pinned connection is prohibited (for this request) then
     // the initiator must decide whether it is OK to open a new one instead.
+    if (!usePinned)
+        request->pinnedConnection()->pinning.peerAccessDenied = true;
```

Comment:
It probably does not matter in the current code, but if feels wrong to make the `true` value sticky. From the PeerSelector point of view, the decision applies to _this_ check only.

```suggestion
        request->pinnedConnection()->pinning.peerAccessDenied = !usePinned;
```

It is kind of wrong to stuff decision-scoped peerAccessDenied into long-term pinnedConnection()->pinning, but I am guessing that it is difficult to relay this flag via PeerSelector result.

Reply:
Fixed in commit e131bf4

> It is kind of wrong to stuff decision-scoped peerAccessDenied into long-term pinnedConnection()->pinning, but I am guessing that it is difficult to relay this flag via PeerSelector result.

Currently the PeerSelector returns a simple list of Comm::Connection objects. We need to update the PeerSelector to return a Comm::Connection/Details pair for this.